### PR TITLE
Add custom steam grid images for imported games

### PR DESCRIPTION
**Pull request by FusRoDah061**
I took a shot at #4 and used SteamGridDb (https://www.steamgriddb.com) service with UWP to provide grid images to imported games. If there are other services like this out there, I'd be happy to know.

What it does is send 4 requests per game to the SteamGridDb API in order to fetch all image variants: hero, logo, grid vertical and horizontal. The first result of each response will be the image to save.

There's a new setting to store SteamGridDb's API key:
image

The images won't be downloaded if no API key is provided.

At the library, it should look like any other Steam game:
uwp_grid_library
uwp_grid_library_hover
uwp_grid_game

### CODE REVIEW

**Comment by FusRoDah061** [comment]
I noticed some of the changes made at #44 might break this.

To generate the game id to use on the images name, steam uses the `exe` entry of the vdf file. Since the app AUMID was removed from the executable path, the exe path will always be the same, which might result in the computed crc beign always the same as well.

I'll have to test this later to be sure, though.

**Comment by FusRoDah061** [comment]
> I noticed some of the changes made at #44 might break this.
> 
> To generate the game id to use on the images name, steam uses the `exe` entry of the vdf file. Since the app AUMID was removed from the executable path, the exe path will always be the same, which might result in the computed crc beign always the same as well.
> 
> I'll have to test this later to be sure, though.

Just tested this, everything's working fine. I missed this line on the `GenerateSteamGridAppId` method:

`byte[] nameTargetBytes = Encoding.UTF8.GetBytes(appTarget + appName + "");`

Concatenating the `appName` with the `appTarget` (exe path), the checksum will remain unique.

**Comment by BrianLima** [comment]
Just tested it, everything works beautifully, my biggest problem was the CRC32 code, thanks for the PR! I'll be approving soon

**Comment by BrianLima** [comment]
Would there be a problem if i let a key pre-registered in the app? Would simplify the usability a bit!

File: UWPHook/Crc32.cs
Comment (BrianLima) [inline UWPHook/Crc32.cs:3]: Would you be willing to add a few comments to this file explaining how the bit wise operations are working here and a little bit of the logic? If you translated the code from another project, a comment stating where does this algorithm came from would be great!

File: UWPHook/GamesWindow.xaml.cs
Comment (BrianLima) [inline UWPHook/GamesWindow.xaml.cs:121]: I really like the usage of asyncs here, just to note 👍

Comment (BrianLima) [reply UWPHook/GamesWindow.xaml.cs:121]: You can resolve this.

Comment (BrianLima) [reply UWPHook/GamesWindow.xaml.cs:121]: I will rework the main load method to async in the next release.

File: UWPHook/GamesWindow.xaml.cs
Comment (BrianLima) [inline UWPHook/GamesWindow.xaml.cs:253]: So, while the grid images are stored on a user basis, both the CRC and the images themselves are the same for any given user using Steam in the same computer.

We are re-downloading images and re-calculating the CRC for any given game * the number of Steam users detected in the system.

We could improve this a bit by moving a bit of the decision and downloading here to the beginning of the ExportGames task, so that we only download and calculate these once, we may store them in a temp folder for a moment, then copy the images over to each user.

Maybe the crc calculation isn't too much of an overhead, but at least the image download would be cool to make a bit less redundant, since many users don't have fast internet and this may take a while.

What do you think?

File: UWPHook/SteamGridDb/SteamGridDbApi.cs
Comment (BrianLima) [inline UWPHook/SteamGridDb/SteamGridDbApi.cs:118]: Maybe would be cool to parameterize these, offering a user choice, like the Playnite launcher offers, many people who use SteamGridDB would appreciate this.

Comment (BrianLima) [reply UWPHook/SteamGridDb/SteamGridDbApi.cs:118]: But you may resolve this and, once merged, i could add these settings!

**Review by BrianLima — CHANGES_REQUESTED** [review]
Very nice overall, my only mandatory request is the redundant image download, everything else i will work on once merged, thanks for the contributions!!! 😊🎉

**Comment by FusRoDah061** [comment]
> Would there be a problem if i let a key pre-registered in the app? Would simplify the usability a bit!

I agree. I don't think there's any problem in doing that.

To get the key you need to sign in to steamgriddb through steam, but their API currently doesn't expose any user info.

File: UWPHook/Crc32.cs
Comment (FusRoDah061) [reply UWPHook/Crc32.cs:3]: I got this code from [this](https://github.com/nullfx/NullFX.CRC) nuget package, actually (at the time I didn't see it). I try replacing this class with said package.

File: UWPHook/SteamGridDb/SteamGridDbApi.cs
Comment (FusRoDah061) [reply UWPHook/SteamGridDb/SteamGridDbApi.cs:118]: I'll leave this up to you then, since I'm not familiar with Playnite, so you'll be able to organize it better, imo

File: UWPHook/GamesWindow.xaml.cs
Comment (FusRoDah061) [reply UWPHook/GamesWindow.xaml.cs:253]: I agree. I didn't notice this behaviour before. I'll work on this.

**Comment by BrianLima** [comment]
> 
> 
> > Would there be a problem if i let a key pre-registered in the app? Would simplify the usability a bit!
> 
> I agree. I don't think there's any problem in doing that.
> 
> To get the key you need to sign in to steamgriddb through steam, but their API currently doesn't expose any user info.

Nice to know! Unless there is some rate-limiting i see no problem with this, probably saving the key alongside the app in a json or something for sake of simplicity, i will contact the SteamGridDB team and ask for authorization to do this!

File: UWPHook/Crc32.cs
Comment (FusRoDah061) [reply UWPHook/Crc32.cs]: Resolved in [e3abf49](https://github.com/BrianLima/UWPHook/pull/50/commits/e3abf499ba304fb45e1d02316df1495bb83a3b4b)

The package I mentioned earlier is .NET Core exclusive, so I went with [Crc32.NET](https://www.nuget.org/packages/Crc32.NET/).

File: UWPHook/GamesWindow.xaml.cs
Comment (FusRoDah061) [reply UWPHook/GamesWindow.xaml.cs]: Made the changes on [22577c8](https://github.com/BrianLima/UWPHook/pull/50/commits/22577c84ae54dec1497850bfe2619d2386b48dc2)

I did as you suggested: download all images to a temp folder, then after the users loop, copy the images to each user grid directory.

It's much faster now!

**Review by BrianLima — APPROVED** [review]

**Pull request merged**
